### PR TITLE
[pytree] align `__all__` for C++ and Python pytree

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112278
* #112111
* __->__ #112110

cc @zou3519 @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi

### CODE REVIEW

File: test/export/test_export.py [hunk 1]
```diff
@@ -670,16 +670,23 @@ class MyDataClass:
         roundtrip_spec = treespec_loads(treespec_dumps(spec))
         self.assertEqual(roundtrip_spec, spec)
 
+        @dataclass
+        class MyOtherDataClass:  # the pytree registration don't allow registering the same class twice
+            x: int
+            y: int
+            z: int = None
+
```
Comment (zou3519) [inline test/export/test_export.py:678]: Did you mean to put this in the next PR up when we unify python and C++ pytree registration?

Comment (XuehaiPan) [reply test/export/test_export.py]: I have moved this to the next PR.

File: torch/utils/_cxx_pytree.py [hunk 2]
```diff
@@ -26,8 +26,16 @@
...
+from ._pytree import TreeSpec as PyTreeSpec
+
+try:
+    import optree
+except ModuleNotFoundError:
+    optree = None  # type: ignore[assignment]
+else:
+    from optree import (  # type: ignore[assignment]  # noqa: F811
+        PyTreeSpec,  # direct import for type annotations
+    )
```
[the remark this answers is not in the document]
Comment (XuehaiPan) [reply torch/utils/_cxx_pytree.py:38]: > that errors gracefully when optree isn't available

I do this by referring to https://github.com/pytorch/pytorch/pull/93139#issuecomment-1433609862.

> Make it an optional dependency, and have a fallback to the python pytree when optree isn't installed.

I think raising an error would be better when it is not available. Currently, we do not have a public pytree API, so we do not need to handle the fallback mechanism right now.

[the remark this answers is not in the document]
Comment (XuehaiPan) [reply torch/utils/_cxx_pytree.py]: I have update the PR.

File: torch/utils/_cxx_pytree.py [hunk 3]
```diff
@@ -187,14 +195,6 @@ def register_pytree_node(
             )
         )
     """
-    from ._pytree import _register_pytree_node
-
-    _register_pytree_node(
-        cls,
-        flatten_func,
-        unflatten_func,
-    )
```
Comment (XuehaiPan) [inline torch/utils/_cxx_pytree.py:196]: Moved to the next PR.

File: torch/utils/_cxx_pytree.py [hunk 4]
```diff
@@ -32,7 +32,12 @@
 
 __all__ = [
     "PyTree",
-    "PyTreeSpec",
+    "Context",
+    "FlattenFunc",
+    "UnflattenFunc",
+    "TreeSpec",
+    "LeafSpec",
+    "_register_pytree_node",
```
Comment (Skylion007) [inline torch/utils/_cxx_pytree.py:40]: Why are we exporting an underscore function?

**Review by Skylion007 — COMMENTED** [review]
Why do we export a presumably private function?

File: torch/utils/_cxx_pytree.py (hunk 4 above)
Comment (XuehaiPan) [reply torch/utils/_cxx_pytree.py:40]: Previously we did not have an `__all__` list in the `torch.utils._pytree` module. But the function `torch.utils._pytree._register_pytree_node` is already used in many downstream packages. I will remove the private APIs in the `__all__` list.

File: torch/utils/_cxx_pytree.py [hunk 5]
```diff
@@ -32,7 +32,11 @@
 
 __all__ = [
     "PyTree",
-    "PyTreeSpec",
+    "Context",
+    "FlattenFunc",
+    "UnflattenFunc",
```
Comment (zou3519) [inline torch/utils/_cxx_pytree.py:37]: Can we add a test that the `__all__` matches? Otherwise they will easily diverge

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**